stl/inc/memory

```diff
@@ -2416,26 +2437,298 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
+    mutable atomic<uintptr_t> _Refptr = 0;
+};
+
+template <class _Ty>
+struct atomic<shared_ptr<_Ty>> : _Atomic_ptr_base<_Ty> {
+private:
+    using _Base = _Atomic_ptr_base<_Ty>;
+
+public:
+    using value_type = shared_ptr<_Ty>;
+
+    static constexpr bool is_always_lock_free = false;
+
+    _NODISCARD bool is_lock_free() const noexcept {
+        return false;
+    }
+
+    void store(const shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_store_memory_order(_Order);
+        _Value._Incref();
+        _Ref_count_base* _Ref = this->_LockAndExchange(_Value._Rep);
+        if (_Ref) {
+            _Ref->_Decref();
+        }
+        this->_Ptr = _Value._Ptr;
+        this->_Unlock();
+    }
+
+    _NODISCARD shared_ptr<_Ty> load(const memory_order _Order = memory_order::seq_cst) const noexcept {
+        _Check_load_memory_order(_Order);
+        shared_ptr<_Ty> _Result;
+        _Ref_count_base* _Ref = this->_LockAndLoad();
+        _Result._Ptr          = this->_Ptr;
+        _Result._Rep          = _Ref;
+        _Result._Incref();
+        this->_Unlock();
+        return _Result;
+    }
+
+    operator shared_ptr<_Ty>() const noexcept {
+        return load();
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(
+        const shared_ptr<_Ty> _Value, const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);
+        _Value._Incref();
+        shared_ptr<_Ty> _Result;
+        _Ref_count_base* _Ref = this->_LockAndExchange(_Value._Rep);
+        _Result._Ptr          = this->_Ptr;
+        _Result._Rep          = _Ref;
+        this->_Ptr            = _Value._Ptr;
+        this->_Unlock();
+        return _Result;
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Order);
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, const shared_ptr<_Ty> _Desired,
+        const memory_order _Order = memory_order::seq_cst) noexcept {
+        _Check_memory_order(_Order);
+        _Ref_count_base* _Ref = this->_LockAndLoad();
+        if (this->_Ptr == _Expected._Ptr && _Ref == _Expected._Rep) {
+            this->_Ptr = _Desired._Ptr;

```

Comment:
These blocks should look like my proposed way to do load/store/exchange above,

Reply:
Are they OK now?